captum/optim/_core/optimization.py

```diff
@@ -56,15 +56,17 @@ def __init__(
                         optimization.
             lr (float): The learning rate to use with the Adam optimizer.
         """
-        self.model = model
+        self.model = model or nn.Identity()
         # Grab targets from loss_function
-        if hasattr(loss_function.target, "__iter__"):
+        assert hasattr(loss_function, "target")
+        if isinstance(loss_function.target, list):
             self.hooks = ModuleOutputsHook(loss_function.target)
         else:
             self.hooks = ModuleOutputsHook([loss_function.target])
         self.input_param = input_param or NaturalImage((224, 224))
         if isinstance(self.model, Iterable):
-            self.input_param.to(next(self.model.parameters()).device)
+            if list(self.model.parameters()) != []:

```

Comment:
covert the generator to a list means we have to iterate the entire generator and store them, which is unnecessarily expensive. If you just want to check if the model has parameters, try call `next` once with fallback
```py
param = next(self.model.parameters(), None)
if param:
  self.input_param.to(param.device)
```

Reply:
@aobo-y Thanks for the tip! I've changed the code to your suggested less intensive way of checking for parameters to use as a source for what device to use.